refactor(docusarus): docusaurus migration

MatanBobi opened the pull request:
This is a WIP PR, some pages still don't work and I still didn't merge all changes from master due to some conflicts.

Open issues:
- [X] Build is broken
- [X] No sidebars on netlify preview
- [ ] Fix CSS for both Dark Mode and Light mode + Tweak the colors to match the current colors
- [x] Update Footer to contain all links
- [x] Fix Layout for Help Page
- [x] Markdown isn't rendered in a `GridBlock` component as it used to
- [x] Blog link is missing from navigation
- [x] Extract the `Showcase` component and use it in both Users page and Home page
- [x] Blog sidebar is missing
- [x] Users page isn't working
- [x] Toggle users logos between dark and light mode and add a comment in `docusaurus.config.js` to add both `darkImage` and `lightImage`.
- [x] Use Tabs in MDX when needed
- [x] Update README
- [x] Fix broken links in existing markdown pages
- [ ] Add `@docusaurus/plugin-pwa` for PWA and offline support
- [x] After all fixes, merge new docs from master
- [ ] Review to see all docs are working and change them to MDX

Feel free to comment here with any problems you see in the preview that weren't mentioned above.

This PR will close https://github.com/testing-library/testing-library-docs/issues/487

MatanBobi commented:
@nickmccurdy following your message [here](https://github.com/testing-library/testing-library-docs/issues/487#issuecomment-689744469), as you can see since I didn't want to break the folder structure, we still have a `website` folder (ignore the `website-old`, this is just for me to compare some stuff).
Do you think we should pull the site to the root folder?

MatanBobi commented:
Looks like the build is broken, I'm on it.

MatanBobi commented:
No sidebar at the moment, I'm checking why it's there locally but not in the netlify preview.

alexkrolick commented on docs/example-external.md:
```diff
@@ -20,5 +20,5 @@ sidebar_label: External Examples
 
 <a href="https://youtu.be/JKOwJUM4_RM">
-  <img width="200px" alt="what is react testing library" src='https://img.youtube.com/vi/JKOwJUM4_RM/0.jpg' style="margin-left: 36px" />
+  <img width="200px" alt="what is react testing library" src='https://img.youtube.com/vi/JKOwJUM4_RM/0.jpg' style={{marginLeft: 36}} />
```
What's up with this? Are the MD files being parsed as MDX now? This syntax is only valid for React.

MatanBobi replied:
Yep, docusaurus 2 uses mdx, I will rename all md later on..

MatanBobi replied:
I've added a task for this above so I'm resolving this. Thanks!

alexkrolick commented on website/src/components/Container/index.js:
```diff
@@ -0,0 +1,42 @@
+/**
```
Why are these layout components being added? Are they no longer included in Docusaurus by default?

MatanBobi replied:
Yeah, they have been removed from v2 so I've temporarily added them, I hope I won't need them or I'll write them differently.

alexkrolick commented on website/src/css/custom.css:
```diff
@@ -0,0 +1,2005 @@
+:root {
```
This looks like the entire vendor file - is there a way to keep the default base stylesheet separate from our customizations? Otherwise, it will be difficult to update when the base file changes.

MatanBobi replied:
Yeah, I haven't finished the CSS stuff, it's the next thing I'm planning to tackle :)

MatanBobi commented:
@kentcdodds @alexkrolick 
Wdyt we should do in dark mode with the `user` section? Most logos look bad in dark mode.

kentcdodds commented:
Maybe we can make sure that only those logos that look good in dark mode make it in the homepage. The others can just make the user's page look bad 😅🤷‍♂️

MatanBobi commented:
> Maybe we can make sure that only those logos that look good in dark mode make it in the homepage. The others can just make the user's page look bad 😅🤷‍♂️

This will leave us with something like 6 in the main page 😅

kentcdodds commented:
I think having fewer on the homeowner is fine actually. And asking for a light and dark themed version of the logo would probably be good too

alexkrolick commented:
What do you think about deferring the work on the light/dark theme until everything else works, especially if we need to change the logos? I'd also skip the other nice-to-haves like PWA support since we aren't sure if those are super useful and/or a maintenance issue going forward.

kentcdodds commented:
100% agreed

MatanBobi commented:
Got it, I'll start working on the other stuff then.

alexkrolick commented on docs/dom-testing-library/api-events.mdx:
````diff
@@ -138,23 +138,31 @@ test that a callback passed to the function was called, or what it was called
-  expect(handleClick).toHaveBeenCalledTimes(1)
-})
-```
+import Tabs from '@theme/Tabs';
````
Nit: maybe we should put imports at the top of the page?

MatanBobi replied:
Definitely, done.

MatanBobi commented:
@alexkrolick @kentcdodds I merged all docs from master and updated them to MDX.
For some reason I still have conflicts even though I merged everything.. Since I don't have write access here, I'm not sure why there are conflicts because I've fixed these specific files locally..
Can one of you please help with that?

MatanBobi commented:
@kentcdodds @alexkrolick this is basically ready, just waiting for your decision regarding the blog sidebar.
Once we decide on that, I'll merge the docs again :)

Thanks!

kentcdodds commented:
Let's do what's easiest for now and we can iterate later. :+1:

alexkrolick commented:
I noticed a few final issues to address before we merge:

1. There seems to be a broken "Introduction" item in the Frameworks menu.
![image](https://user-images.githubusercontent.com/1571667/97224570-91922100-178e-11eb-936e-ff889bcb437f.png)

2. The Recipes and Docs top bar items can be highlighted at the same time

<img width="517" alt="Screen Shot 2020-10-26 at 1 25 27 PM" src="https://user-images.githubusercontent.com/1571667/97224761-d7e78000-178e-11eb-82b8-989741c3d40e.png">

3. The primary color for links is not really legible in dark mode

<img width="851" alt="Screen Shot 2020-10-26 at 1 26 45 PM" src="https://user-images.githubusercontent.com/1571667/97224849-f6e61200-178e-11eb-89f7-2b5a730adf21.png">

MatanBobi commented:
> I noticed a few final issues to address before we merge:
> 
> 1. There seems to be a broken "Introduction" item in the Frameworks menu.
> ![image](https://user-images.githubusercontent.com/1571667/97224570-91922100-178e-11eb-936e-ff889bcb437f.png)
> 
> 2. The Recipes and Docs top bar items can be highlighted at the same time
> 
> <img width="517" alt="Screen Shot 2020-10-26 at 1 25 27 PM" src="https://user-images.githubusercontent.com/1571667/97224761-d7e78000-178e-11eb-82b8-989741c3d40e.png">
> 
> 3. The primary color for links is not really legible in dark mode
> 
> <img width="851" alt="Screen Shot 2020-10-26 at 1 26 45 PM" src="https://user-images.githubusercontent.com/1571667/97224849-f6e61200-178e-11eb-89f7-2b5a730adf21.png">
> 
> 

Thanks @alexkrolick ! Will fix, run sanity and ping here.

MatanBobi commented:
Hi @alexkrolick, I fixed all the issues and also found a way to add the blog sidebar :)
For the link color in dark mode, I've just set it to a lighter one but I'm not a UX expert so I hope it will look better.
I reviewed and things look good besides the images of the users on dark mode, I had a fix for that but then there was a change that randomizes the users on the front page so it got overriden, not sure it's that important though.
If you think it's ready for merge, I'll merge the latest docs (I want to do it as close as possible to the merge because it takes some time).

Another 2 questions, can we pull out the `.vscode` folder from the repo and add it to `.gitignore`? I'm not aware of any reason to keep it there.
Second, do we use the docker files somewhere? `netlify.toml` isn't using it and I didn't change the files but things still work properly..

alexkrolick commented:
> If you think it's ready for merge, I'll merge the latest docs (I want to do it as close as possible to the merge because it takes some time).

Looks good to me, let's do it

> can we pull out the .vscode folder from the repo and add it to .gitignore? I'm not aware of any reason to keep it there.

Actually there is useful for stuff for working on the docs in there. Can you update the code snippet for code blocks to the MDX way of doing it, unless it is super obvious (in which case you can delete)? https://github.com/testing-library/testing-library-docs/blob/master/.vscode/markdown.code-snippets

> Second, do we use the docker files somewhere? netlify.toml isn't using them and I didn't change the files but things still work properly..

This came with the Docusaurus template but if it's not used by Netlify we probably don't need it

MatanBobi commented:
Done @alexkrolick :)
Actually was cool to learn about adding the code snippets to vscode and I think that the tabs snippet can be helpful since it has some indentation and line breaks tricks.
I'm converting this from a draft PR to a PR.

nickmccurdy commented:
> @nickmccurdy following your message [here](https://github.com/testing-library/testing-library-docs/issues/487#issuecomment-689744469), as you can see since I didn't want to break the folder structure, we still have a `website` folder (ignore the `website-old`, this is just for me to compare some stuff).
> Do you think we should pull the site to the root folder?

@MatanBobi Yes, since that seems to be the convention for Docusaurus 2.

MatanBobi commented:
> > @nickmccurdy following your message [here](https://github.com/testing-library/testing-library-docs/issues/487#issuecomment-689744469), as you can see since I didn't want to break the folder structure, we still have a `website` folder (ignore the `website-old`, this is just for me to compare some stuff).
> > Do you think we should pull the site to the root folder?
> 
> @MatanBobi Yes, since that seems to be the convention for Docusaurus 2.

@nickmccurdy Done :)

alexkrolick commented:
@MatanBobi functionally, looks good.

What happened to the stripes? It's not critical, but looks a bit empty now layout-wise

## Before

<img width="1499" alt="Screen Shot 2020-11-03 at 12 40 40 PM" src="https://user-images.githubusercontent.com/1571667/98038339-d0008f00-1dd1-11eb-865c-296638c28444.png">

## After
<img width="1889" alt="Screen Shot 2020-11-03 at 12 39 36 PM" src="https://user-images.githubusercontent.com/1571667/98038270-b3fced80-1dd1-11eb-8aa4-54ea70106dae.png">

MatanBobi commented:
@alexkrolick I removed those because they didn't fit the styling in dark mode..
I can play with that a little bit though to make a dark mode compatible "light" background

alexkrolick commented:
Fantastic. Merging.

nickmccurdy commented:
@all-contributors add @alexkrolick for code

nickmccurdy commented:
@all-contributors add @MatanBobi  for code
